gb-apu/src/apu.rs

```diff
@@ -0,0 +1,179 @@
+                        (*ve).step();
+                    }
+                }
+            }
+        }
+
+        if self.cycle_counter % 0x5F == 0 {
+            self.add_sample();
+        }
+        if self.buffer_i >= BUFFER_SIZE {
+            self.queue_audio();
+            self.buffer_i = 0;
+        }
+    }
+}
+
+impl<A> FileOperation<A, IORegArea> for Apu
+where
+    A: Address<IORegArea>,
+    u16: From<A>,
+{
+    fn read(&self, addr: A) -> Result<u8, Error> {
+        use IORegArea::{
+            Nr10, Nr11, Nr12, Nr13, Nr14, Nr21, Nr22, Nr23, Nr24, Nr30, Nr31, Nr32, Nr33, Nr34,
+            Nr41, Nr42, Nr43, Nr44, Nr50, Nr51, Nr52,
+        };
+        match addr.area_type() {
+            Nr10 | Nr11 | Nr12 | Nr13 | Nr14 => self.sound_channels[0].read(addr),
+            Nr21 | Nr22 | Nr23 | Nr24 => self.sound_channels[1].read(addr),
+            Nr30 | Nr31 | Nr32 | Nr33 | Nr34 => self.sound_channels[2].read(addr),
+            Nr41 | Nr42 | Nr43 | Nr44 => self.sound_channels[3].read(addr),
+            Nr52 => Ok(if self.enabled { 0x80 } else { 0 }
+                | if self.sound_channels[3].enabled {
+                    0x8
+                } else {
+                    0
+                }
+                | if self.sound_channels[2].enabled {
+                    0x4
+                } else {
+                    0
+                }
+                | if self.sound_channels[1].enabled {
+                    0x2
+                } else {
+                    0
+                }
+                | if self.sound_channels[0].enabled {
+                    0x1
+                } else {
+                    0
+                }),
+            _ => Err(Error::SegmentationFault(addr.into())),
+        }
+    }
+    fn write(&mut self, v: u8, addr: A) -> Result<(), Error> {
+        use IORegArea::{
+            Nr10, Nr11, Nr12, Nr13, Nr14, Nr21, Nr22, Nr23, Nr24, Nr30, Nr31, Nr32, Nr33, Nr34,
+            Nr41, Nr42, Nr43, Nr44, Nr50, Nr51, Nr52,
+        };
+        match addr.area_type() {
+            Nr10 | Nr11 | Nr12 | Nr13 | Nr14 => return self.sound_channels[0].write(v, addr),
+            Nr21 | Nr22 | Nr23 | Nr24 => return self.sound_channels[1].write(v, addr),
+            Nr30 | Nr31 | Nr32 | Nr33 | Nr34 => return self.sound_channels[2].write(v, addr),
+            Nr41 | Nr42 | Nr43 | Nr44 => return self.sound_channels[3].write(v, addr),
+            Nr52 => {
+                let was_enabled = self.enabled;
+                let enabled = v & 0x80 != 0x00;
+                if was_enabled && !enabled {
+                    self.sound_channels = vec![
+                        SoundChannel::new(ChannelType::SquareWave, true),
+                        SoundChannel::new(ChannelType::SquareWave, false),
+                        SoundChannel::new(ChannelType::WaveForm, false),
+                        SoundChannel::new(ChannelType::Noise, false),
+                    ];
```

Comment:
Are we required to recreate all the channel when we re-enable the sound ?

Reply:
All the registers Nrx0..Nrx4 must be reset, so I figured it was simpler to create new channels since all their variables are mirroring the registers

Reply:
I'm worried about recreating a new vector, but It don't seems to cause a lot of problems RN

Reply:
I don't think disabling the apu disable is used a lot anyway, but it's worth testing it